doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+These actions are mediated by the system, and take effect only when the system confirms the action.
+Users can defer and batch these actions, modulo lease timeouts, imposing only on the freshness and efficiency of the system.
+
+Capabilities may expire, as a lease, which may result in errors from attempts to exercise the capability.
+
+It is very common for users to immediately drop capabilities if they do not require the ability.
+The write capability for most TVCs is dropped immediately (by the system) unless it is meant to be written to by users (e.g. a `TABLE`).
+The read capability may be dropped immediately if the user does not presently need to read the data.
+Capabilities may be recovered by the best-effort command `AcquireCapabilities(id)` to be discussed.
+
+## Storage
+
+The Storage layer presents a narrow API about which it makes guarantees.
+It likely has a more verbose diagnostic API that describes its state, which should be a view over the results of these commands.
+
+*   `Create(id, description) -> (ReadCapability, WriteCapability)`: binds to `id` the TVC described by `description`.
+
+    The command returns capabilities naming `id`, with frontiers set to initial values chosen by the Storage layer.
+
+    The collection associated with `id` is based on `description`, but it is up to the Storage layer to determine what the source's contents will be.
+    A standarad example is a Kafka topic, whose contents are "added" to the underlying collection as they are observed.
+    The Storage layer only needs to ensure that it can reliably produce the updates between `since` and `upper`, and otherwise the nature of the content is up for discussion.
+    It is an error to re-use a previously used `id`.
+
+*   `SubscribeAt(ReadCapability(id, frontier))`: returns a snapshot of `id` at `frontier`, followed by an ongoing stream of subsequent updates.
+
+    This command returns the contents of `id` as of `frontier` once they are known, and updates thereafter once they are known.
+    The snapshot and stream contain in-line statements of the subscription's `upper` frontier: those times for which updates may still arrive.
+    This call may block, or not be promptly responded to, if `frontier` is greater or equal to the current `upper` of `id`.
+    The subscription can be canceled by either endpoint, and the recipient should only downgrade their read capability when they are certain they have all data for an interval.
+
+    A subscription can be constructed with additional arguments that change how the data is returned to the user.
+    For example, the user may ask to not receive the initial snapshot, rather that receive and then discard it.
+    For example, the user may provide filtering and projection that can be applied before the data are transmitted.
+
+*   `UpdateAndDowngrade(WriteCapability(id, frontier), updates, new_frontier)`: applies `updates` to `id` and downgrades `frontier` to `new_frontier`.
```

Comment:
I had a hard time reasoning about what "durability" meant otherwise. Like, if you can write at your write capability, without advancing the write capability, what does it mean for that to be durable given that it can change before the frontier ticks and it commits. If you try to write some stuff, crash, come back up, and want to know "did I write it?" you'd probably have a hard time.

What seems easier to reason about is when you have a write capability, writing and downgrading at the same time. If you want to have multiple write capabilities, because you want to write lots of things at related times, that works for me. But, by advancing a capability as you write you leave it unambiguous whether the writes associated with that capability landed or not.

Reply:
Ah, that all makes sense. I think perhaps what is missing is the text that asserts that new_frontier is strictly greater than the current frontier. I had assumed you could pass in the current frontier as the new frontier to retain the capability to write more data at the current time.

Reply:
Ah, oops I'll add that. You could technically call this with the frontiers equal and `updates` empty, but if you used a non-empty `updates` it would be an error (update times need to be beyond on frontier and not beyond the next frontier).